### ISSUE TITLE
[Feature proposal] Lightning Address for LNURLp/wallet

### ISSUE THREAD

**Issue by Darth-Coin** [lnbits/lnbits#1035]
### Description
Using LNbits as a "bank" for family and friends, sometimes you want also to offer them a LN address.
I know it exist the extension "Lightning Address" that lets you rent LN addresses from other domains. I see that for a more "commercial" use.

### Necessity
I would like also to have the option to set a LN address on each LNbits wallet / LNURL-p for users, so when I setup an account for each, I could set a specific LN address for each wallet or LNURL-p of that wallet.
LNbits will get to be used more and more and is needed easy interfaces to set these things. I know there's a way to do it manually by adding a json txt file with specs for LNUJRL-p, but that is quite advanced for many users and require knowledge and instructions.
Let's make it easy to use for anybody, straight from LNbits UI.

### Solution
Something like this, in the LNURL-p setup screen, add a field for LN address designated to that user.
![image](https://user-images.githubusercontent.com/56997124/194032876-01abbf0b-be01-4f31-86a2-15d16c77eef6.png)

**Comment by talvasconcelos** [comment]
This could be easily achievable in a WoS style... 

reference: https://github.com/williexu/random_username/blob/master/random_username/generate.py

**Comment by bitkarrot** [comment]
This can work for the lnurlp extension, assuming several things: 

1. the username added in the lnurlp extension will auto create at the domain where you host your lnbits instance. For example, if your lnbits is running on https://lnbits.site.com, then all lightning addresses created will be in this format: username@lnbits.site.com 
2. all usernames will be lowercase and follow the rule set for lightning addresses will be applied in the new Lnurlp extention form field and the end user will be informed. (Note: email addresses can have upper and '_' but not lightning addresses)
3. Assume uvicorn can serve up the .well-known directory like a regular website.
4. Adjustments to be made to forms, database and API in the lnurlp extension. 

I'd like to take a stab and self assign this task as it will be useful to have the feature on the lnbits server instead of having to run 'bridge' code in order to pair LNBits lnurlps with usernames for laisee.org. Adding the lightning address feature directly to the lnurlp extension could eliminate this extra step. 

Note: This proposed addition to Lnurlp extension does not make it possible to host the lightning addresses at a domain different from where the lnbits instance is hosted. 

See original [LNURL-pay Specification](https://github.com/lnurl/luds/blob/legacy/lnurl-pay.md)

**Comment by bitkarrot** [comment]
I forgot to mention that we also have to consider the other lightning extension addresses that are created, and potential conflict there (if any)

**Comment by bitkarrot** [comment]
Looks like this is already implemented for "Lightning Addresses" Extension for renting. Will need some way to also handle for Free option in lnurlp.

```
@core_app.get("/.well-known/lnurlp/{username}")
async def lnaddress(username: str, request: Request):
    from lnbits.extensions.lnaddress.lnurl import lnurl_response

    domain = urlparse(str(request.url)).netloc
    return await lnurl_response(username, domain, request)
```

Please some suggestion from the LNBits maintainers on how best you would like to handle this if we also include Free Lightning Address inside of LNURLp extension. e.g.  add `from lnbits.extensions.lnurlp.lnurl import lnurl_response` 

A switch could be one option. Or perhaps check if extension is activated? 
@motorina0 ? Thanks in advance :) 


https://github.com/lnbits/lnbits/blob/4f7e9119e876bb559c9a328e05efdd550a471cdc/lnbits/core/views/public_api.py#L18

**Comment by motorina0** [comment]
> Looks like this is already implemented for "Lightning Addresses" Extension for renting. Will need some way to also handle for Free option in lnurlp.

I believe this PR https://github.com/lnbits/lnbits/pull/1060 should solve it. Reviewing it now.

**Comment by motorina0** [comment]
One option is for the `amin` users to be allowed to manually add addresses and map the addresses to a wallet.
@Darth-Coin @bitkarrot what do you think?

![image](https://user-images.githubusercontent.com/2951406/203026204-e7799fd1-57d9-4482-a9aa-544d2b88c5b9.png)

**Comment by Darth-Coin** [comment]
Yes, this could work nice too, I like it.

**Comment by bitkarrot** [comment]
If Admin only access is required per lightning address, (in order to prevent users from adding additional ones), I can see that being a plus, but it should probably be an flexible option and not a fixed one in any extension.

Most importantly, what admin/invoice keys are required to generate the LN Addresses from the API?  If the user is generated by User Manager, possibly limit if they can or cannot add additional lnaddresses. Every LNBits instance will likely have a different community rule set.

I am thinking of the use case where you can give a user 1 free LNaddress but then additional addresses would have a cost to it.  Ideally we don't need 1 sat charge for the free address, and have some sort of flexible ruleset. Unclear if this is implemented or is in the plan to be.

Anyway, maybe I am misunderstanding what you are trying to say, do let me know :) 
 

> One option is for the `amin` users to be allowed to manually add addresses and map the addresses to a wallet. @Darth-Coin @bitkarrot what do you think?

**Comment by bezysoftware** [comment]
This would be great. I just spend several hours trying to figure out why my lightning address setup wasn't working ("description hash doesn't match") and the issue was that LNBits constructed the metadata like this:
`[["text/plain", "Pay to abcd"]]`

But I had this in my well-known file:
`[["text/plain","Pay to abcd"]]`

The difference was a missing space, which obviously caused two different `description_hash`es. 

Having this built into LNBits would solve this. Or at the very least it could generate the well-known json, so I can just copy-paste it and be sure I don't make a typo

**Comment by tlindi** [comment]
I want to have a way to restrict random users to access this satdress "create Address" page.
![image](https://user-images.githubusercontent.com/7249767/207183126-883b3252-083b-4186-bf3b-cc4626b73da9.png)
FYI: above link example is not online.

We want to sell Lightning Addresses for domains, and I do not want random user to be able to access a registration page of "satdress". But satdress should still work as an (endpoint?) for generating invoice(?)

I have nginx proxy in front of server, so workaround could be some URL filter for nginx conf (or so - but I heard it's not recommended).

**Comment by dergigi** [comment]
That would be awesome to have, even if it's just `random-generated-name@lnbits.site.com` for every (lnurlp-enabled) wallet.

**Comment by bitkarrot** [comment]
Hi, I'm posting this as per request by @dergigi 

This is the Free Lightning Address Extension as described by @Darth-Coin and  https://github.com/lnbits/lnbits/issues/1035#issuecomment-1320721645

Because the LNBits main line has massive changes in the last few days,  I am sharing a working branch and also a separate repository. 

If you install is **older than Jan 3, 2023,** 
Follow the steps in this repository
https://github.com/bitkarrot/free-lnaddress-lnbits

If your install is currently up to date with the LNBits main branch **(  > Jan 4, 2023 )** try my current branch and be sure to edit your .env accordingly (see above repo for install notes)
https://github.com/bitkarrot/lnbits-legend/tree/free-lnaddress

If you have feedback or find bugs please make an issue on my repo. This extension was a bit of a hack for my own projects, but since people keep asking about it, I am posting it here.

**Comment by tlindi** [comment]
Might you be able to make this as not conflicting extension, but add functionality to existing Lightning Address extension as a new feature/switch?

**Comment by bitkarrot** [comment]
> @BitCarrot Might you be able to make this as not conflicting extension, but add functionality to existing Lightning Address extension as a new feature/switch?

See https://github.com/lnbits/lnbits/issues/1035#issuecomment-1321812586

**Comment by bitkarrot** [comment]
I will make PR (as requested by Ben Arc) of below branch  to mainline after fixing some bugs that were discovered today.

> If your install is currently up to date with the LNBits main branch **( > Jan 4, 2023 )** try my current branch and be sure to edit your .env accordingly (see above repo for install notes) https://github.com/bitkarrot/lnbits-legend/tree/free-lnaddress

**Comment by bitkarrot** [comment]
We merged this into lnurlp about a month ago. Can we now close this issue?

**Comment by Darth-Coin** [comment]
Fixed in #1319
